### PR TITLE
No nullability warnings in best type and type inference

### PR DESCRIPTION
**Pull request by jcouv**
The invariant merge rules no longer produce a warning when there is a mismatch.
So scenarios like `new[] { x, y }` and `M(x, y)` should not report any diagnostics for failing to find a best type including nullability. Instead, we find a best type, and then produce conversion warnings from the elements to the inferred type.
Then elements can be suppressed.
This also affects the conditional operator.

This allows removing much machinery for passing `hadNullabilityMismatch` around, and a few "no best type" error codes.

Fixes https://github.com/dotnet/roslyn/issues/30376 (suppression in array initializer and conditional)
Fixes https://github.com/dotnet/roslyn/issues/30151 (suppression in array initializer)
Fixes https://github.com/dotnet/roslyn/issues/30480 (no nullability warning for type inference with lambda argument)

I'll push an additional commit shortly which:
Closes https://github.com/dotnet/roslyn/issues/30955
Closes https://github.com/dotnet/roslyn/issues/30925
Closes https://github.com/dotnet/roslyn/issues/31958
Closes https://github.com/dotnet/roslyn/issues/30941
Closes https://github.com/dotnet/roslyn/issues/28377
Closes https://github.com/dotnet/roslyn/issues/31295
Closes https://github.com/dotnet/roslyn/issues/26654
(Note: I verified that those issues aren't referenced elsewhere in the code)

### CODE REVIEW

**Comment by jcouv** [comment]
@gafter @dotnet/roslyn-compiler for review. Details up top. Thanks

File: src/Compilers/CSharp/Portable/BoundTree/BoundExpressionWithNullability.cs [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed partial class BoundExpressionWithNullability : BoundExpression
+    {
+        public override bool IsSuppressed
+        {
+            get => Expression.IsSuppressed;
+            protected set => throw ExceptionUtilities.Unreachable;
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/BoundTree/BoundExpressionWithNullability.cs]: Should we mark the setter as `[Obsolete]`?

Comment (jaredpar) [reply src/Compilers/CSharp/Portable/BoundTree/BoundExpressionWithNullability.cs]: Hmm I'm not sure that would really work here. 

---
In reply to: [252448228](https://github.com/dotnet/roslyn/pull/32947#discussion_r252448228) [](ancestors = 252448228)

File: src/Compilers/CSharp/Portable/BoundTree/BoundNode.cs [hunk 2]
```diff
@@ -158,7 +158,7 @@ public void ResetCompilerGenerated(bool newCompilerGenerated)
             }
         }
 
-        public bool IsSuppressed
+        public virtual bool IsSuppressed
```
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNode.cs]: Did you consider the approach where you just had `BoundExperssionWithNullability` set this in the ctor based on the value of `Expression.IsSuppressed`? That could avoid the use of `virtual` here as long as we were confident that `Expression` would never be set after the ctor.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNode.cs]: I did, but the constructor is generated.

Comment (jcouv) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNode.cs]: I could add a constructor overload that takes a `bool isSuppressed` flag. Do you think that's better?

Comment (jaredpar) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNode.cs]: We could but that would also be eays to miss in code. Should we just remove this ctor from the generator and hand author it? I believe there is some precedence for that. 

---
In reply to: [252452436](https://github.com/dotnet/roslyn/pull/32947#discussion_r252452436) [](ancestors = 252452436)

Comment (jcouv) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNode.cs]: We don't currently have a way of not generating constructors. But I created an overload without `hasErrors`.

File: docs/features/nullable-reference-types.md [hunk 3]
```diff
@@ -236,16 +236,12 @@ It is intended that these merging rules are associative and commutative, so that
 > ***Open issue***: these rules do not describe the handling of merging pointer types.
 
 ### Array creation
-The calculation of the _best type_ element nullability uses the Conversions rules above.
-The top-level and nested nullability are calculated independently.
-The top-level nullability is the most relaxed of the elements, where `!` is a `~` is a `?`.
-The nested nullability is the merged nullability of the best common type. If there is a merge conflict,
-the nested nullability is `~` and a warning is reported.
+The calculation of the _best type_ element nullability uses the Conversions rules above and the contravariant merging rules.
```
Comment (cston) [inline docs/features/nullable-reference-types.md]: >contravariant [](start = 96, length = 13)

"covariant"?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 4]
```diff
@@ -727,11 +727,16 @@ private void ReportAssignmentWarnings(BoundExpression value, TypeSymbolWithAnnot
                 var destinationType = targetType.TypeSymbol;
                 if ((object)sourceType != null && IsNullabilityMismatch(destinationType, sourceType))
                 {
-                    ReportSafetyDiagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, value.Syntax, sourceType, destinationType);
+                    ReportNullabilityMismtachInAssignment(value.Syntax, sourceType, destinationType);
                 }
             }
         }
 
+        private void ReportNullabilityMismtachInAssignment(SyntaxNode syntaxNode, object sourceType, object destinationType)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs]: >Mismtach [](start = 38, length = 8)

Typo.

**Comment by cston** [comment]
>         public void SuppressNullableWarning_Conditional2()

Why rename? The operations are null-coalescing.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:36500 in 284d282. [](commit_id = 284d282d8d039c2082fc2885882e8640a14505c5, deletion_comment = False)

**Comment by jcouv** [comment]
> Why rename? 
Was a mistake. Added a test with a conflicting name, then renamed the wrong one. Thanks

**Pull request merged**